Fixes #10713 - keep more information in the backtrace in production environment

iNecas opened the pull request:
1. in the full trace in UI, keep the backtrace from plugins, otherwise, the
   errors from the plugin pages are without useful inforamtion
2. in the production.log, log the backtrace at error level: asking the user to
   switch to the debug first is not very user friendly: we should try to get as much
   needed info as possible from the beginning.

domcleal commented on app/controllers/application_controller.rb:
```diff
@@ -296,7 +296,7 @@ def saved_redirect_url_or(default)
   def generic_exception(exception)
     logger.warn "Operation FAILED: #{exception}"
-    logger.debug exception.backtrace.join("\n")
+    logger.error exception.backtrace.join("\n")
```
While I'm inclined to agree it's a problematic policy (having backtraces at debug level) we have a lot of this throughout the codebase.  If it's to be changed, it would have to be everywhere.

It should also be at the same level as the exception message itself, else it would look rather strange if the log level was set at `error`.

domcleal replied:
A few more thoughts...
1. If it's at the same log level, we can log it as a single multi-line message so they don't become separated.
2. We could run this backtrace through the silencer, making it shorter (but thanks to your other changes, still accurate).  If debug's enabled, then it could be non-silenced and longer?
3. Combine this into one easy to use helper :)

iNecas replied:
I can do it everywhere the grep find it: should not be that hard

domcleal replied:
Yeah, they should all be pretty similar.

iNecas replied:
Updated: I kept 2 commits there: one for the silencers, the second for the logging of backtraces

domcleal commented on config/initializers/backtrace_silencers.rb:
```diff
@@ -1,6 +1,10 @@
 # You can add backtrace silencers for libraries that you're using but don't wish to see in your backtraces.
+
+# keep the lines from plugins in the backtrace
+Foreman::Plugin.modify_backtrace_cleaner
```
Could we keep the silencer modifications in this file?  It seems like a more appropriate home than Foreman::Plugin itself - perhaps just keep a helper there if necessary.

iNecas replied:
Good point: I ended up in this layout mainly as evolution of the solution (I tried several ways, where the first versions needed bigger interaction with the plugin class), updated

domcleal commented on app/services/foreman/plugin.rb:
```diff
@@ -67,6 +67,16 @@ def register(id, &block)
+        cleaner.remove_silencers!
+        cleaner.add_silencer do |line|
+          line !~ Rails::BacktraceCleaner::APP_DIRS_PATTERN &&
+              !(all.any? { |plugin| line.include?(plugin.name) })
```
I think we also have plugin.root available for pointing to the installation dir, which might be more precise.

iNecas replied:
unfortunatelly, the rails add the filters to the backtrace, to replace the full path with the more readable prefix, containing just the name and version instead of full path (https://github.com/rails/rails/blob/master/railties/lib/rails/backtrace_cleaner.rb#L28), so the lines we match look like this:

```
foreman_setup (2.1.1) app/controllers/foreman_setup/provisioners_controller.rb:175:in `block in step4_update'
```

domcleal replied:
Should we replace the filter?  Since line.include?(plugin.name) will only likely work when plugins are namespaced, which isn't all that many - and usually, not every file.

iNecas committed:
Fixes #10713 - improved backtrace logging

* full trace in 500 page in production shows really the full trace
* Foreman::Logging.exception as an unified entry point for exceptions logging
* the backtrace cleaner doesn't delete the plugins
* the orchestration exceptions gets the backtrace covered by logger
* some generic resuced exceptions get covered by logger

domcleal commented on app/controllers/application_controller.rb:
```diff
@@ -107,7 +107,7 @@ def api_deprecation_error(exception = nil)
     process_error(:redirect => :back, :error_msg => exception.message)
     logger.warn "ProxyAPI operation FAILED: #{exception}"
-    logger.debug exception.backtrace.join("\n")
+    logger.warn exception.backtrace.join("\n")
```
From earlier comment:

In these instances, we should probably log it as a single entry now they're both at the same log level - so we know they'll always both be logged, they could be one multi-line log message.

If we're always printing backtraces, then we should probably put them through the backtrace silencer - at least, if !debug mode(?).  Put this in a common helper?

iNecas replied:
is the added value of sparing few lines in the log file big enough to loose additional data for investigation? For example the backtrace from fog might be really important

domcleal replied:
Though this PR still configures and uses the silencer - so you intend to show the silenced stack in the UI, but the full stack in the logs?  Should the UI show a silenced stack and a full one - or even just the full one, and remove all silencing?

iNecas replied:
The more I think about it the more I think the silencer makes more sense in development, where usually the exceptions mean problems with your current changes (so the silencer makes it easier to find the issues quickly as you write the code), however in production, they are throwing away useful information. I'm ok with introducing a helper to log the exception: in that case I would revisit also the logging in orchestration, where we often throw away the backtrace https://github.com/theforeman/foreman/blob/6ace062d9dd9f0a529639d8123bf778072dce18e/app/models/concerns/orchestration/realm.rb#L17, so much for small PRs :)

domcleal replied:
I agree!

> so much for small PRs :)

Ah, but when we probe to find the deep, underlying issues, we end up with something so much more valuable :)

iNecas replied:
Finding good name for the helper: what about `Foreman.log_exception(e)`

domcleal replied:
Yeah, or under Foreman::Logging, e.g. `Foreman::Logging.exception(e, level)`?

domcleal replied:
Er, though that doesn't make it clear what the destination logger is.

iNecas replied:
what about `options` as the second argument, accepting optionally:
- logger
- description: such as "Unhandled VMware error", that we have on couple of places
- level: warn by default

domcleal replied:
That could work.  I'd probably make the description mandatory though so people don't skimp, I can't think of a reason not to include it.  (And probably call it `message`.)

I was looking at logging, and [it seemed that its loggers even supported writing out backtraces](https://github.com/TwP/logging/blob/master/lib/logging/logger.rb#L23).. but it doesn't seem to accept an exception as the second argument for me.

iNecas replied:
Isn't the exception.message descriptive enough in many cases?

domcleal replied:
Adding a message is usually pretty important for context - otherwise you'd have to trace back from the log file to figure out what called it.  All of the instances you've changed in this PR have some message that adds context.

iNecas replied:
I'm afraid people will pass the `exception.message` as the `message` too often

iNecas replied:
In that case I would suggest: `Foreman::Logging.exception("Context message", e, options)` with optional options `level` and `logger`

ares replied:
Could we make it a bit shorter? `Foreman::Log.exception` maybe? And one small comment, I'd like to have `e` as the first argument, context might be long message and the exception could be hard to find. Otherwise looking forward to have this available in Foreman.

domcleal replied:
@iNecas sounds fine to me.

@ares the reason we're considering Foreman::Logging was that we already have such a module, so I'm not sure we should create one with such a similar name.

ares replied:
Ah ok, makes sense.

iNecas commented:
I've introduced the Foreman::Logging.exception helper and converted couple of places where we throw away the backtrace (mostly the orchestration). We still do so on couple of places, but the problem of those is some exceptions are expected while us not being specific enough on the type of exception we count on. I don't think it's in human abilities to fix all places at once, but would be nice to focus on occurrence of `generic rescue` in the PR process.

iNecas commented:
I've also made the 'full trace' in production to really show full trace, instead of the cleaned version. Left the cleaner there mainly for the developement environment

ares commented on lib/foreman/logging.rb:
```diff
@@ -52,6 +52,23 @@ def logger(name)
+      logger_name = options[:logger] || 'app'
+      level       = options[:level] || :warn
+      unless ::Logging::LEVELS.keys.include?(level.to_s)
+        raise "Unexpected level #{level}, expected one of #{ ::Logging::LEVELS.keys }"
```
maybe one nit `Unexpected log level`

domcleal commented on app/controllers/hosts_controller.rb:
```diff
@@ -677,10 +678,10 @@ def find_multiple
-    logger.debug e.message
-    logger.debug e.backtrace.join("\n")
+  rescue => error
+    message = _("Something went wrong while selecting hosts - %s") % e
```
`e` or `error`, but not both

iNecas replied:
Good catch, fixed

domcleal commented on app/controllers/dashboard_controller.rb:
```diff
@@ -4,6 +4,7 @@ class DashboardController < ApplicationController
   before_filter :find_resource, :only => [:destroy]
 
   def index
+    raise 'hello'
```
hi!

iNecas replied:
LOL :D sorry for that

ares replied:
- [x] check reviewers attention

domcleal commented on app/controllers/application_controller.rb:
```diff
@@ -5,7 +5,7 @@ class ApplicationController < ActionController::Base
   ensure_security_headers
   protect_from_forgery # See ActionController::RequestForgeryProtection for details
   rescue_from ScopedSearch::QueryNotSupported, :with => :invalid_search_query
-  rescue_from Exception, :with => :generic_exception if Rails.env.production?
```
Also an accident?

I'm doing precisely the same here...

domcleal commented:
@iNecas thanks, please squash into one commit

iNecas commented:
Squashed and rebased

domcleal commented:
Merged as 01e7826, thanks @iNecas.

iNecas commented:
Thanks @domcleal
